tests/qnn/test_qnn_torch.py

```diff
@@ -0,0 +1,421 @@
+
+    @pytest.mark.parametrize("dtype", [torch.float32, torch.float64])
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(2))
+    @pytest.mark.parametrize("batch_size", [2])
+    def test_step_module(self, module, batch_size, n_qubits, output_dim, dtype):
+        """Test if a module that includes TorchLayers can perform one optimization step. This
+        test checks that some of the parameters in the module are different after one step.
+        The module is composed of two TorchLayers sandwiched between Linear neural network layers,
+        and the dataset is simply input and output vectors of zeros."""
+        module = module.type(dtype)
+        loss_func = torch.nn.MSELoss()
+        optimizer = torch.optim.SGD(module.parameters(), lr=0.5)
+
+        x = torch.zeros((batch_size, n_qubits)).type(dtype)
+        y = torch.zeros((batch_size, output_dim)).type(dtype)
+
+        params_before = [w.detach().numpy().copy() for w in list(module.parameters())]
+
+        module_out = module(x)
+        optimizer.zero_grad()
+        loss = loss_func(module_out, y)
+        loss.backward()
+        optimizer.step()
+
+        params_after = [w.detach().numpy().copy() for w in list(module.parameters())]
+
+        params_similar = [np.allclose(p1, p2) for p1, p2 in zip(params_before, params_after)]
+        assert not all(params_similar)
+
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(2))
+    def test_module_gradients(self, module, output_dim, n_qubits, get_circuit):
+        """Test if a gradient can be calculated with respect to all of the trainable variables in
+        the module"""
+        c, w = get_circuit
+
+        x = torch.zeros((2, n_qubits))
+        y = torch.zeros((2, output_dim))
+
+        module_out = module(x)
+        loss_func = torch.nn.MSELoss()
+        loss = loss_func(module_out, y)
+        loss.backward()
+
+        gradients = [w.grad for w in module.parameters()]
+        assert all([g.is_floating_point() for g in gradients])
+        assert len(gradients) == 2 * len(w) + 6  # six parameters come from classical layers
+
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    def test_module_state_dict(self, module, output_dim, n_qubits, get_circuit):
+        """Test if the state dictionary output by the module contains all the expected trainable
+        parameters"""
+        c, w = get_circuit
+
+        state_dict = module.state_dict()
+        dict_keys = set(state_dict.keys())
+
+        clayer_weights = set("clayer{}.weight".format(i + 1) for i in range(3))
+        clayer_biases = set("clayer{}.bias".format(i + 1) for i in range(3))
+        qlayer_params = set(
+            "qlayer{}.w{}".format(i + 1, j + 1) for i in range(2) for j in range(len(w))
+        )
+
+        all_params = clayer_weights | clayer_biases | qlayer_params
+
+        assert dict_keys == all_params
+        assert len(dict_keys) == len(all_params)
```

Comment:
Tests look good, but just wondering if there are any edge cases (weird output/input for QNodes) that we might be missing

Reply:
In terms of input I think we've specified quite well in the docstring:
![image](https://user-images.githubusercontent.com/49409390/81223089-6be04800-8fb3-11ea-9604-9885fc5ec523.png)
I can think of edge cases where the second point is not satisfied, e.g., if someone puts a list of arrays as a single argument. We don't test for this case, do you think we need to? I think it's not needed since we say not to do this in the docstring and the user's `weight_shapes` dictionary also wouldn't have made sense in this case.

In terms of output, I think you pointed out one case where the QNode returns multiple arrays of different size. Is this something that's supported in PL generally, as I got an error as I mention in the comment thread above (https://github.com/XanaduAI/pennylane/pull/588/files#r420086869).

Reply:
Yep, I think the docstring does quite a good job!